### ISSUE TITLE
Transforming expressions

### ISSUE THREAD

**Issue by danieljharvey** [danieljharvey/horses#480]
We could do a bunch of useful optimisations, particularly before generating JS.

Ideas for now:

## Float lets up

Given:

```haskell
\a -> \b -> let a2 = a + 100; b + a2
```

We are recalculating `a + 100` every time the function is run. We'd change this to:

```haskell
\a -> let a2 = a + 100; \b -> b + a2
```

Therefore the `a + 100` only happens once if the second part of the function is called multiple times.

### Also float them down:

Given:

```haskell
let a = expensiveOperation 100;
match Just 1 with
  (Just aa) -> aa + a
  Nothing -> 0
```

We'd change this to

```haskell
match Just 1 with
  (Just aa) -> let a = expensiveOperation; aa + a
  Nothing -> let a = expensiveOperation; 0
```

The second `let a = expensiveOperation` would then be binned by the unused rule above.

**Comment by danieljharvey** [comment]
One important thing to do here is make sure that external deps are also removed when no longer used.

**Comment by danieljharvey** [comment]
Unused vars implemented in #479
